src/data/nat/choose/multinomial.lean

```diff
@@ -117,3 +129,107 @@ lemma multinomial_univ_three (a b c : ℕ) : multinomial finset.univ ![a, b, c]
 by simp [multinomial, fin.sum_univ_three, fin.prod_univ_three]
 
 end nat
+
+/-! ### Alternative definitions -/
+
+namespace finsupp
+
+variables {α : Type*}
+
+/-- Alternative multinomial definition based on a finsupp, using the support
+  for the big operations
+-/
+def multinomial (f : α →₀ ℕ) : ℕ := (f.sum $ λ _, id)! / f.prod (λ _ n, n!)
+
+lemma multinomial_eq (f : α →₀ ℕ) : f.multinomial = nat.multinomial f.support f := rfl
+
+lemma multinomial_update (a : α) (f : α →₀ ℕ) :
+  f.multinomial = (f.sum $ λ _, id).choose (f a) * (f.update a 0).multinomial :=
+begin
+  simp only [multinomial_eq],
+  classical,
+  by_cases a ∈ f.support,
+  { rw [← finset.insert_erase h, nat.multinomial_insert _ f (finset.not_mem_erase a _),
+      finset.add_sum_erase _ f h, support_update_zero], congr' 1,
+    exact nat.multinomial_congr _
+      (λ _ h, (function.update_noteq (finset.mem_erase.1 h).1 0 f).symm) },
+  rw not_mem_support_iff at h,
+  rw [h, nat.choose_zero_right, one_mul, ← h, update_self],
+end
+
+end finsupp
+
+namespace multiset
+
+variables {α : Type*}
+
+/-- Alternative definition of multinomial based on `multiset` delegating to the
+  finsupp definition
+-/
+noncomputable def multinomial (m : multiset α) : ℕ := m.to_finsupp.multinomial
```

Comment:
Can you add the `rfl` lemmas that connects this definition with the previous ones?

Reply:
In `multinomial_filter_ne` below, I can already `rw multinomial` (or `simp_rw multinomial` to rewrite both), so I think the lemma is already there (auto-generated equational lemma?). Do you want a `rfl` lemma that directly expands to `nat.multinomial m.finsupp.support m.finsupp`?

Reply:
If `rw` works it's enough, thanks!